tests/test_types.py

```diff
@@ -60,3 +60,25 @@ def test_build(self):
         room = RoomAlias("channel", "my.domain")
 
         self.assertEquals(room.to_string(), "#channel:my.domain")
+
+
+class GroupIDTestCase(unittest.TestCase):
+    def test_parse(self):
+        group_id = GroupID.from_string("+group/=_-.123:my.domain")
+        self.assertEqual("group/=_-.123", group_id.localpart)
+        self.assertEqual("my.domain", group_id.domain)
+
+    def test_validate(self):
+        bad_ids = [
+            "$badsigil:domain",
+            "+:empty",
+        ] + [
+            "+group" + c + ":domain" for c in "A%?æ£"
+        ]
+        for id_string in bad_ids:
+            try:
+                GroupID.from_string(id_string)
+                self.fail("Parsing '%s' should raise exception" % id_string)
+            except SynapseError as exc:
+                self.assertEqual(400, exc.code)
+                self.assertEqual("M_UNKNOWN", exc.errcode)
```

Comment:
FWIW I think you should be able to use [`self.assertRaises`](http://twistedmatrix.com/documents/current/api/twisted.trial._synctest._Assertions.html#assertRaises):

```python
with self.assertRaises(SynapseError) as exc:
    GroupID.from_string(id_string)

self.assertEqual(400, exc.code)
self.assertEqual("M_UNKNOWN", exc.errcode)
```

Though actually I guess that is no cleaner at all than just try...except

Reply:
yeah. I did it this way so that I could show which id_string was failing to cause an exception.